logging/src/logging.cpp

```diff
@@ -0,0 +1,432 @@
+    }
+
+    void
+    operator()(dynamic_t::null_t) {
+        throw std::system_error(std::make_error_code(std::errc::invalid_argument), "can not process null as attribute value");
+    }
+
+    template<class T>
+    void
+    operator()(const T& value) {
+        attributes.emplace_back(blackhole::attribute_t(name, value));
+    }
+
+    blackhole::attributes_t& attributes;
+    const std::string& name;
+    typedef void result_type;
+};
+
+void
+logging_t::deleter_t::operator()(impl_t* ptr) {
+    delete ptr;
+};
+
+const io::basic_dispatch_t&
+logging_t::prototype() const {
+    return *this;
+}
+
+logging_t::logging_t(context_t& context, asio::io_service& asio, const std::string& service_name, const dynamic_t& service_args) :
+    service_t(context, asio, service_name, service_args),
+    dispatch<io::base_log_tag>(service_name),
+    impl(new impl_t(
+        context,
+        [&]() {
+            auto registry = blackhole::registry_t::configured();
+            registry.add<blackhole::sink::console_t>();
+
+            std::stringstream stream;
+            stream << boost::lexical_cast<std::string>(context.config.logging.loggers);
+
+            return registry.builder<blackhole::config::json_t>(stream).build(service_args.as_object().at("backend", "core").as_string());
+        }(),
+        service_args.as_object().at("unicorn_path", "/logging_v2/filters").as_string()
+    ))
+{
+    on<io::base_log::get>(std::make_shared<logging_slot_t>(*this));
+    on<io::base_log::list_loggers>(std::bind(&impl_t::list_loggers, impl.get()));
+    on<io::base_log::set_filter>(std::bind(&impl_t::set_filter, impl.get(), ph::_1, ph::_2, ph::_3, logging::filter_t::disposition_t::local));
+    on<io::base_log::remove_filter>(std::bind(&impl_t::remove_filter, impl.get(), ph::_1));
+    on<io::base_log::list_filters>(std::bind(&impl_t::list_filters, impl.get()));
+    on<io::base_log::set_cluster_filter>(std::bind(&impl_t::set_filter, impl.get(), ph::_1, ph::_2, ph::_3, logging::filter_t::disposition_t::cluster));
+}
+
+named_logging_t::named_logging_t(logging::logger_t& _log, std::string _name, std::shared_ptr<logging::metafilter_t> _filter) :
+    dispatch<io::named_log_tag>(std::move(_name)),
+    log(_log),
+    filter(std::move(_filter))
+{
+    on<io::named_log::emit>(std::bind(&named_logging_t::emit, this, ph::_1, ph::_2, ph::_3));
+    on<io::named_log::emit_ack>(std::bind(&named_logging_t::emit_ack, this, ph::_1, ph::_2, ph::_3));
+}
+
+void
+named_logging_t::emit(const std::string& message, unsigned int severity, const logging::attributes_t& attributes) {
+    emit_ack(message, severity, attributes);
+}
+
+bool
+named_logging_t::emit_ack(const std::string& message, unsigned int severity, const logging::attributes_t& attributes) {
+    if(filter->apply(message, severity, attributes) == logging::filter_result_t::reject) {
+        return false;
+    }
+    if(!attributes.empty()) {
+        blackhole::attributes_t bh_attributes;
```

Comment:
So instead of reusing the code you initially convert from bytes into heap allocated objects which in its turn are converted into another *possibly* heap allocated objects, right? And all that mess is required just to not depend from blackhole in public API?